File: #6. 2022 KAKAO TECH INTERNSHIP/박정현/problem1_성격유형검사하기.java
// 풀이 시간 : 25분
// 시간 복잡도: O(N)
// 코드 설명: 모든 경우를 하드 코딩 했읍니다...... 몇개 안되더라구요........
// 취약한 부분: 아무래도 있어빌리티가 좀 떨어지는 감이 없지않아 있다.

import java.util.*;

class Solution {
    public String solution(String[] survey, int[] choices) {
        String answer = "";
        int[] arr = new int[4];
        String[] types = {"RT", "CF", "JM", "AN"};
        for(int i=0;i<survey.length;i++){
            switch(survey[i]){
                case "RT":{
                    arr[0] += choices[i] - 4;
                    break;
                }
                case "TR":{
                    arr[0] += 4 - choices[i];
                    break;
                }
                case "CF":{
                    arr[1] += choices[i] - 4;
                    break;
                }
                case "FC":{
                    arr[1] += 4 - choices[i];
                    break;
                }
                case "JM":{
                    arr[2] += choices[i] - 4;
                    break;
                }
                case "MJ":{
                    arr[2] += 4 - choices[i];
                    break;
                }
                case "AN":{
                    arr[3] += choices[i] - 4;
                    break;
                }
                case "NA":{
                    arr[3] += 4 - choices[i];
                    break;
                }
            }
        }
        System.out.println(Arrays.toString(arr));
        for(int i=0;i<4;i++){
            if (arr[i] <= 0) answer += types[i].charAt(0);
            else answer += types[i].charAt(1);
        }
        return answer;
    }
}

File: #6. 2022 KAKAO TECH INTERNSHIP/장효빈/problem1_성격_유형_검사하기.py
# 풀이 시간: 20분
# 시간 복잡도: O(N)
# 코드 설명: 성격 유형을 키로 해서 점수를 더해줌
# 취약한 부분: 어떻게 하면 깔끔할까 고민

def solution(survey, choices):
    d = {'R': 0, 'T': 0, 'C': 0, 'F': 0, 'J': 0, 'M': 0, 
         'A': 0, 'N': 0}
    
    for s, c in zip(survey, choices):
        if c < 4:
            d[s[0]] += 4 - c
        elif c > 4:
            d[s[1]] += c % 4
            
    answer = ''
    for key in ('RT', 'CF', 'JM', 'AN'):
        if d[key[0]] >= d[key[1]]:
            answer += key[0]
        else:
            answer += key[1]
    
    return answer